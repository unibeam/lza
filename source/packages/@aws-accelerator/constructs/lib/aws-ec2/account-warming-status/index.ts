/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
/**
 * aws-ec2-pre-warm-account-status - lambda handler
 *
 * @param event
 * @returns
 */

import { SSMClient, PutParameterCommand } from '@aws-sdk/client-ssm';
import {
  EC2Client,
  DescribeInstancesCommand,
  DescribeVpcsCommand,
  DescribeSubnetsCommand,
  DeleteVpcCommand,
  DeleteSubnetCommand,
  TerminateInstancesCommand,
} from '@aws-sdk/client-ec2';
import { AdaptiveRetryStrategy } from '@aws-sdk/util-retry';
import { delay } from '@aws-accelerator/utils/lib/throttle';

const solutionId = process.env['SOLUTION_ID'] ?? '';
const retryStrategy = new AdaptiveRetryStrategy(() => Promise.resolve(5));
const ec2Client = new EC2Client({ customUserAgent: solutionId, retryStrategy });
const ssmClient = new SSMClient({ customUserAgent: solutionId, retryStrategy });

type InstanceDetails = {
  instanceId: string | undefined;
  minutesSinceLaunch: number | undefined;
};

//eslint-disable-next-line @typescript-eslint/no-explicit-any
export async function handler(event: any): Promise<
  | {
      IsComplete: boolean;
    }
  | undefined
> {
  console.log(event);
  const ssmPrefix: string = event.ResourceProperties['ssmPrefix'];
  const instanceDetails = await getInstanceDetails();

  // if no instance was found, it has been deleted return
  if (instanceDetails.instanceId === undefined) {
    return { IsComplete: true };
  }

  if (instanceDetails.minutesSinceLaunch! > 15) {
    console.log('Account warming time reached');
    await terminateInstance(instanceDetails.instanceId);
    await deleteVpc();
    await setSSMParameter(ssmPrefix, 'true');
    return {
      IsComplete: true,
    };
  }
  console.log('Waiting for warmup to complete');
  return { IsComplete: false };
}

async function getInstanceDetails(): Promise<InstanceDetails> {
  console.log('Getting Instance Id');
  const instanceDetails: InstanceDetails = { instanceId: undefined, minutesSinceLaunch: undefined };
  const ec2Instances = await ec2Client.send(
    new DescribeInstancesCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (ec2Instances.Reservations && ec2Instances.Reservations.length < 1) {
    return instanceDetails;
  }

  if (ec2Instances.Reservations![0].Instances?.length ?? 0 > 0) {
    instanceDetails.instanceId = ec2Instances.Reservations![0].Instances![0].InstanceId;
    const launchTime = ec2Instances.Reservations![0].Instances![0].LaunchTime;

    const currentTime = new Date();
    const msElapsed = currentTime.getTime() - launchTime!.getTime();
    instanceDetails.minutesSinceLaunch = msElapsed / 1000 / 60;
  }

  return instanceDetails;
}

async function deleteVpc() {
  console.log('Deleting VPC');
  const vpcId = await getVpcId();

  if (vpcId) {
    console.log('Deleting subnets');
    const subnets = await ec2Client.send(
      new DescribeSubnetsCommand({ Filters: [{ Name: 'vpc-id', Values: [vpcId] }] }),
    );
    for (const subnet of subnets.Subnets) {
      await ec2Client.send(new DeleteSubnetCommand({ SubnetId: subnet.SubnetId }));
    }
    console.log(`Deleting VPC with id: ${vpcId}`);
    await ec2Client.send(new DeleteVpcCommand({ VpcId: vpcId }));
  }
}

async function getVpcId(): Promise<string | undefined> {
  console.log('Getting VPC Id');
  const vpcs = await ec2Client.send(
    new DescribeVpcsCommand({ Filters: [{ Name: 'tag:Name', Values: ['accelerator-warm'] }] }),
  );

  if (vpcs.Vpcs?.length ?? 0 > 0) {
    return vpcs.Vpcs![0].VpcId;
  }
  return undefined;
}

async function setSSMParameter(ssmPrefix: string, parameterValue: string) {
  console.log('Updating SSM Parameter');
  try {
    await ssmClient.send(
      new PutParameterCommand({
        Name: `${ssmPrefix}/account/pre-warmed`,
        Value: parameterValue,
        Overwrite: true,
      }),
    );
  } catch (e) {
    console.log(e);
  }
}

async function terminateInstance(instanceId: string): Promise<void> {
  console.log(`Terminating EC2 instance: ${instanceId}`);
  await ec2Client.send(new TerminateInstancesCommand({ InstanceIds: [instanceId] }));
  await waitForTermination(instanceId);
}

async function waitForTermination(instanceId: string): Promise<void> {
  console.log(`Waiting for termination of instanceId: ${instanceId}`);
  let ec2Terminated = false;
  while (!ec2Terminated) {
    const statusResponse = await ec2Client.send(new DescribeInstancesCommand({ InstanceIds: [instanceId] }));
    if (
      statusResponse.Reservations[0].Instances[0].State?.Name !== 'terminated' ||
      statusResponse.Reservations[0].Instances[0].State?.Code !== 48
    ) {
      delay(15000);
      continue;
    }
    ec2Terminated = true;
  }
}
