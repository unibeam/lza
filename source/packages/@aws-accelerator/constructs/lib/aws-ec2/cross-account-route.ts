/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as path from 'path';

interface CrossAccountRouteFrameworkProps {
  /**
   * Accelerator Prefix
   */
  readonly acceleratorPrefix: string;
  /**
   * Custom resource lambda log group encryption key, when undefined default AWS managed key will be used
   */
  readonly logGroupKmsKey?: cdk.aws_kms.IKey;

  /**
   * Custom resource lambda log retention in days
   */
  readonly logRetentionInDays: number;
}

export class CrossAccountRouteFramework extends cdk.Resource {
  public readonly provider: cdk.custom_resources.Provider;

  constructor(scope: Construct, id: string, props: CrossAccountRouteFrameworkProps) {
    super(scope, id);

    const onEventStsPolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'StsAssumeRole',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['sts:AssumeRole'],
      resources: [`arn:${cdk.Stack.of(this).partition}:iam::*:role/${props.acceleratorPrefix}*`],
    });

    const onEventRoutePolicy = new cdk.aws_iam.PolicyStatement({
      sid: 'EC2RouteCreateDelete',
      effect: cdk.aws_iam.Effect.ALLOW,
      actions: ['ec2:CreateRoute', 'ec2:DeleteRoute'],
      resources: ['*'],
    });

    const onEvent = new cdk.aws_lambda.Function(this, 'CrossAccountRouteFunction', {
      code: cdk.aws_lambda.Code.fromAsset(path.join(__dirname, 'cross-account-route/dist')),
      runtime: cdk.aws_lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      timeout: cdk.Duration.seconds(15),
      description: 'Cross account EC2 route OnEvent handler',
      initialPolicy: [onEventStsPolicy, onEventRoutePolicy],
    });
    new cdk.aws_logs.LogGroup(this, `${onEvent.node.id}LogGroup`, {
      logGroupName: `/aws/lambda/${onEvent.functionName}`,
      retention: props.logRetentionInDays,
      encryptionKey: props.logGroupKmsKey,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    this.provider = new cdk.custom_resources.Provider(this, 'CrossAccountRouteProvider', {
      onEventHandler: onEvent,
    });
  }
}

interface CrossAccountRouteProps {
  /**
   * The owner account of the route table
   */
  readonly ownerAccount: string;

  /**
   * The region of the owner account
   */
  readonly ownerRegion: string;

  /**
   * The partition of the owner account
   */
  readonly partition: string;

  /**
   * The custom resource provider
   */
  readonly provider: cdk.custom_resources.Provider;

  /**
   * The role name to assume
   */
  readonly roleName: string;

  /**
   * The ID of the route table for the route
   */
  readonly routeTableId: string;

  /**
   * The ID of a carrier gateway
   */
  readonly carrierGatewayId?: string;

  /**
   * The destination CIDR
   */
  readonly destination?: string;

  /**
   * The destination prefix list ID
   */
  readonly destinationPrefixListId?: string;

  /**
   * The ID of an egress-only Internet gateway
   */
  readonly egressOnlyInternetGatewayId?: string;

  /**
   * The ID of an Internet Gateway or Virtual Private Gateway
   */
  readonly gatewayId?: string;

  /**
   * The ID of an instance
   */
  readonly instanceId?: string;

  /**
   * The destination IPv6 CIDR
   */
  readonly ipv6Destination?: string;

  /**
   * The ID of a local gateway
   */
  readonly localGatewayId?: string;

  /**
   * The ID of a NAT gateway
   */
  readonly natGatewayId?: string;

  /**
   * The ID of a network interface
   */
  readonly networkInterfaceId?: string;

  /**
   * The ID of a transit gateway
   */
  readonly transitGatewayId?: string;

  /**
   * The ID of a VPC endpoint
   */
  readonly vpcEndpointId?: string;

  /**
   * The ID of a VPC peering connection
   */
  readonly vpcPeeringConnectionId?: string;
}

export class CrossAccountRoute extends cdk.Resource {
  private roleArn?: string;

  constructor(scope: Construct, id: string, props: CrossAccountRouteProps) {
    super(scope, id);

    const CROSS_ACCOUNT_ROUTE_TYPE = 'Custom::CrossAccountRoute';
    this.roleArn =
      cdk.Stack.of(this).account !== props.ownerAccount
        ? `arn:${props.partition}:iam::${props.ownerAccount}:role/${props.roleName}`
        : undefined;

    new cdk.CustomResource(this, 'Resource', {
      resourceType: CROSS_ACCOUNT_ROUTE_TYPE,
      serviceToken: props.provider.serviceToken,
      properties: {
        region: props.ownerRegion,
        roleArn: this.roleArn,
        routeDefinition: {
          DestinationCidrBlock: props.destination,
          DestinationPrefixListId: props.destinationPrefixListId,
          DestinationIpv6CidrBlock: props.ipv6Destination,
          RouteTableId: props.routeTableId,
          CarrierGatewayId: props.carrierGatewayId,
          EgressOnlyInternetGatewayId: props.egressOnlyInternetGatewayId,
          GatewayId: props.gatewayId,
          InstanceId: props.instanceId,
          LocalGatewayId: props.localGatewayId,
          NatGatewayId: props.natGatewayId,
          NetworkInterfaceId: props.networkInterfaceId,
          TransitGatewayId: props.transitGatewayId,
          VpcEndpointId: props.vpcEndpointId,
          VpcPeeringConnectionId: props.vpcPeeringConnectionId,
        },
      },
    });
  }
}
