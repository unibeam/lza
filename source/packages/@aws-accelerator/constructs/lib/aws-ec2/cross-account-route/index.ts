/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

/**
 * aws-ec2-cross-account-route - lambda handler
 *
 * @param event
 * @returns
 */

import { CloudFormationCustomResourceEvent } from '@aws-accelerator/utils/lib/common-types';
import { CreateRouteCommand, DeleteRouteCommand, EC2Client } from '@aws-sdk/client-ec2';
import { STSClient } from '@aws-sdk/client-sts';
import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import { setRetryStrategy, getStsCredentials } from '@aws-accelerator/utils/lib/common-functions';

export async function handler(event: CloudFormationCustomResourceEvent): Promise<
  | {
      PhysicalResourceId: string;
      Status: string | undefined;
    }
  | undefined
> {
  interface RouteProps {
    readonly RouteTableId: string;
    readonly CarrierGatewayId?: string;
    readonly DestinationCidrBlock?: string;
    readonly DestinationPrefixListId?: string;
    readonly DestinationIpv6CidrBlock?: string;
    readonly EgressOnlyInternetGatewayId?: string;
    readonly GatewayId?: string;
    readonly InstanceId?: string;
    readonly LocalGatewayId?: string;
    readonly NatGatewayId?: string;
    readonly NetworkInterfaceId?: string;
    readonly TransitGatewayId?: string;
    readonly VpcEndpointId?: string;
    readonly VpcPeeringConnectionId?: string;
  }

  const props: RouteProps = event.ResourceProperties['routeDefinition'];
  const region: string = event.ResourceProperties['region'];
  const solutionId = process.env['SOLUTION_ID'];
  const roleArn: string | undefined = event.ResourceProperties['roleArn'];
  const resourceId = setResourceId(props);

  const ec2Client = await setClient(solutionId, region, roleArn);

  switch (event.RequestType) {
    case 'Create':
    case 'Update':
      await throttlingBackOff(() => ec2Client.send(new CreateRouteCommand(props)));

      return {
        PhysicalResourceId: resourceId,
        Status: 'SUCCESS',
      };

    case 'Delete':
      await throttlingBackOff(() =>
        ec2Client.send(
          new DeleteRouteCommand({
            DestinationCidrBlock: props.DestinationCidrBlock,
            DestinationPrefixListId: props.DestinationPrefixListId,
            DestinationIpv6CidrBlock: props.DestinationIpv6CidrBlock,
            RouteTableId: props.RouteTableId,
          }),
        ),
      );

      return {
        PhysicalResourceId: event.PhysicalResourceId,
        Status: 'SUCCESS',
      };
  }

  /**
   * Set physical resource ID based on event input
   * @param props
   * @returns string
   */
  function setResourceId(props: RouteProps): string {
    if (props.DestinationCidrBlock) {
      return `${props.DestinationCidrBlock}${props.RouteTableId}`;
    } else if (props.DestinationIpv6CidrBlock) {
      return `${props.DestinationIpv6CidrBlock}${props.RouteTableId}`;
    } else {
      return `${props.DestinationPrefixListId}${props.RouteTableId}`;
    }
  }

  /**
   * Set EC2 client
   * @param solutionId string | undefined
   * @param region string | undefined
   * @param roleArn string | undefined
   * @returns Promise<EC2Client>
   */
  async function setClient(solutionId?: string, region?: string, roleArn?: string): Promise<EC2Client> {
    if (roleArn) {
      const stsClient = new STSClient({ customUserAgent: solutionId, region, retryStrategy: setRetryStrategy() });

      return new EC2Client({
        customUserAgent: solutionId,
        region,
        retryStrategy: setRetryStrategy(),
        credentials: await getStsCredentials(stsClient, roleArn),
      });
    } else {
      return new EC2Client({ customUserAgent: solutionId, region, retryStrategy: setRetryStrategy() });
    }
  }
}
