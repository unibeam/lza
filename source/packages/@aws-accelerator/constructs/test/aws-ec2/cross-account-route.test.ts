/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';

import { CrossAccountRoute, CrossAccountRouteFramework } from '../../lib/aws-ec2/cross-account-route';
import { snapShotTest } from '../snapshot-test';

const testNamePrefix = 'Construct(CrossAccountRoute): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

const provider = new CrossAccountRouteFramework(stack, 'Framework', {
  logGroupKmsKey: new cdk.aws_kms.Key(stack, 'Key'),
  logRetentionInDays: 3653,
  acceleratorPrefix: 'AWSAccelerator',
}).provider;

new CrossAccountRoute(stack, 'Resourcev4', {
  ownerAccount: 'TestAccount',
  ownerRegion: 'us-east-1',
  partition: 'aws',
  provider,
  roleName: 'TestRole',
  routeTableId: 'rtb-test123',
  destination: '10.0.0.0/16',
  vpcPeeringConnectionId: 'pcx-test123',
});

new CrossAccountRoute(stack, 'Resourcev6', {
  ownerAccount: 'TestAccount',
  ownerRegion: 'us-east-1',
  partition: 'aws',
  provider,
  roleName: 'TestRole',
  routeTableId: 'rtb-test123',
  ipv6Destination: '::/0',
  vpcPeeringConnectionId: 'pcx-test123',
});

new CrossAccountRoute(stack, 'ResourcePl', {
  ownerAccount: 'TestAccount',
  ownerRegion: 'us-east-1',
  partition: 'aws',
  provider,
  roleName: 'TestRole',
  routeTableId: 'rtb-test123',
  destinationPrefixListId: 'pl-test',
  vpcPeeringConnectionId: 'pcx-test123',
});

/**
 * CrossAccountRoute construct test
 */
describe('CrossAccountRoute', () => {
  snapShotTest(testNamePrefix, stack);
});
