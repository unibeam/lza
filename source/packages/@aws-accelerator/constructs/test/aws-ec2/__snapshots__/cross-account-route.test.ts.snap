// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrossAccountRoute Construct(CrossAccountRoute):  Snapshot Test 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Resources": {
    "FrameworkCrossAccountRouteFunction7BE199E2": {
      "DependsOn": [
        "FrameworkCrossAccountRouteFunctionServiceRoleDefaultPolicyEFFA4FAA",
        "FrameworkCrossAccountRouteFunctionServiceRoleD3FBE736",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "Cross account EC2 route OnEvent handler",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "FrameworkCrossAccountRouteFunctionServiceRoleD3FBE736",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 15,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FrameworkCrossAccountRouteFunctionLogGroup9B330ABE": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsKeyId": {
          "Fn::GetAtt": [
            "Key961B73FD",
            "Arn",
          ],
        },
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "FrameworkCrossAccountRouteFunction7BE199E2",
              },
            ],
          ],
        },
        "RetentionInDays": 3653,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "FrameworkCrossAccountRouteFunctionServiceRoleD3FBE736": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FrameworkCrossAccountRouteFunctionServiceRoleDefaultPolicyEFFA4FAA": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":iam::*:role/AWSAccelerator*",
                  ],
                ],
              },
              "Sid": "StsAssumeRole",
            },
            {
              "Action": [
                "ec2:CreateRoute",
                "ec2:DeleteRoute",
              ],
              "Effect": "Allow",
              "Resource": "*",
              "Sid": "EC2RouteCreateDelete",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FrameworkCrossAccountRouteFunctionServiceRoleDefaultPolicyEFFA4FAA",
        "Roles": [
          {
            "Ref": "FrameworkCrossAccountRouteFunctionServiceRoleD3FBE736",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FrameworkCrossAccountRouteProviderframeworkonEvent1A959FA7": {
      "DependsOn": [
        "FrameworkCrossAccountRouteProviderframeworkonEventServiceRoleDefaultPolicyAF7F5043",
        "FrameworkCrossAccountRouteProviderframeworkonEventServiceRole274CC003",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "REPLACED-GENERATED-NAME.zip",
        },
        "Description": "AWS CDK resource provider framework - onEvent (Default/Framework/CrossAccountRouteProvider)",
        "Environment": {
          "Variables": {
            "USER_ON_EVENT_FUNCTION_ARN": {
              "Fn::GetAtt": [
                "FrameworkCrossAccountRouteFunction7BE199E2",
                "Arn",
              ],
            },
          },
        },
        "Handler": "framework.onEvent",
        "Role": {
          "Fn::GetAtt": [
            "FrameworkCrossAccountRouteProviderframeworkonEventServiceRole274CC003",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FrameworkCrossAccountRouteProviderframeworkonEventServiceRole274CC003": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FrameworkCrossAccountRouteProviderframeworkonEventServiceRoleDefaultPolicyAF7F5043": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "FrameworkCrossAccountRouteFunction7BE199E2",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "FrameworkCrossAccountRouteFunction7BE199E2",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FrameworkCrossAccountRouteProviderframeworkonEventServiceRoleDefaultPolicyAF7F5043",
        "Roles": [
          {
            "Ref": "FrameworkCrossAccountRouteProviderframeworkonEventServiceRole274CC003",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Key961B73FD": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "ResourcePl6F481B77": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "FrameworkCrossAccountRouteProviderframeworkonEvent1A959FA7",
            "Arn",
          ],
        },
        "region": "us-east-1",
        "roleArn": "arn:aws:iam::TestAccount:role/TestRole",
        "routeDefinition": {
          "DestinationPrefixListId": "pl-test",
          "RouteTableId": "rtb-test123",
          "VpcPeeringConnectionId": "pcx-test123",
        },
      },
      "Type": "Custom::CrossAccountRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "Resourcev4C9A541C2": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "FrameworkCrossAccountRouteProviderframeworkonEvent1A959FA7",
            "Arn",
          ],
        },
        "region": "us-east-1",
        "roleArn": "arn:aws:iam::TestAccount:role/TestRole",
        "routeDefinition": {
          "DestinationCidrBlock": "10.0.0.0/16",
          "RouteTableId": "rtb-test123",
          "VpcPeeringConnectionId": "pcx-test123",
        },
      },
      "Type": "Custom::CrossAccountRoute",
      "UpdateReplacePolicy": "Delete",
    },
    "Resourcev6A6F941A4": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "FrameworkCrossAccountRouteProviderframeworkonEvent1A959FA7",
            "Arn",
          ],
        },
        "region": "us-east-1",
        "roleArn": "arn:aws:iam::TestAccount:role/TestRole",
        "routeDefinition": {
          "DestinationIpv6CidrBlock": "::/0",
          "RouteTableId": "rtb-test123",
          "VpcPeeringConnectionId": "pcx-test123",
        },
      },
      "Type": "Custom::CrossAccountRoute",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
